### ISSUE TITLE
Fail pages for including reading functions?

### ISSUE THREAD

**Issue by detlevhfischer** [w3c/wcag#1199]
Well-meaning sites often include reading functions supplied by external vendors. If the controls of the reading interface have clear WCAG Failures, does that fail the page, in your opinion? Or would you consider these functions additional 'nice to have' modes that can be treated as inaccessible 'alternative versions' of the otherwise accessible page content?

**Comment by patrickhlauke** [comment]
the controls form part of the page, so fail them. just because they provide access to an alternative presentation/version doesn't exempt them from themselves needing to be accessible

**Comment by patrickhlauke** [comment]
(same way we'd fail, say, video player controls for embedded video content, even if the video itself is essentially an alternative version of text already present in the page)

**Comment by alastc** [comment]
Presumably we wouldn't be failing the audio content (e.g. captions/transcript), but would be failing the keyboard accessibility.

If there was a an inaccessible PDF for print purposes (with the same content as the webpage), I'd consider that an alternative version and wouldn't consider if it had issues. The difference is that it's on a separate page...

**Comment by mraccess77** [comment]
WCAG allows for alternatives on the same page.  An inaccessible graph could have a data table on the same page or linked to from that page to meet WCAG.

**Comment by patrickhlauke** [comment]
but, to turn it to detlev's point, the controls to switch on / switch between the alternatives must be accessible. they're not exempt together with the actual alternative that they provide

**Comment by alastc** [comment]
@mraccess77 agreed, but there is a nuance here. A graph and table are directly equivalent 'content'. 

A video which is supplementary to the page text should still have accessible controls, surely?

How about you have two tabs, one is a map with location pin. The other is text-directions for getting to the location. Do the buttons for the map need to be keyboard accessible?

**Comment by mraccess77** [comment]
> How about you have two tabs, one is a map with location pin. The other is text-directions for getting to the location. Do the buttons for the map need to be keyboard accessible?

That is the question here -- if the buttons are only used to control the alternative and inaccessible content with the primary content being accessible they serve no other purpose.

If you have an inaccessible chart with a table equivalent- but the chart has some sorting - but the table also can be sorted - does the chart sorting feature really need to be accessible because the table sorting feature provides the same functionality?

I'm not fundamentally disagreeing - I'm just trying to make sure we have a clear path to say what is required and what isn't.       So in my example, would the inaccessible chart still need an accessible sorting mechanism even though the table sorting is accessible?

**Comment by patrickhlauke** [comment]
it's probably more about what we mean by inaccessible. is it inaccessible but hidden away? or is, say, a screen reader user going to encounter lots of "button" controls with no accessible name or something, or even a series of focusable controls that have been aria-hidden (so they tab/focus onto nothing), or just "image" with no alternative or anything... i'd fail that sort of stuff.

**Comment by detlevhfischer** [comment]
To add a bit of nuance to my original question: There is just one (well-labelled, keyboard and screen reader accessible) button to bring up the additional read-aloud controls (which then do have issues). One could argue that those who don't need it are not impeded by these issues as they will not activate it - and those who would be most impeded by its flaws are screen reader users who would likely use their screen reader instead of the read-aloud function to consume page content. 
An important audience of the site are people with severe cognitive disabilities, some of whom clearly benefit from the function. That's why it feels so wrong to fail the site for this add-on content.  Another example IMO why the strict PASS/FAIL conformance approach deserves to be re-thought (as is happening in Silver).

**Comment by jake-abma** [comment]
I'd say it's about trust, anxiety and confusion in this case.
Is / are the controls causing these and thus creating an unequal experience between functional needs?

If so (human judgement call?) we only have SC fails for now but probably not as consistent and reliable between testers as we want.
Good point from Detlev to explain this / offer solutions and scoring with future adjectival ratings.

**Comment by detlevhfischer** [comment]
In terms of the current conformance model, I am still not sure whether a read-aloud function should be considered [supplemental content](https://www.w3.org/TR/WCAG21/#dfn-supplementary-content) - since it reproduces content in another mode of use rather than supplementing it - or whether it is acceptable to see it as a non-conformant version of the [conforming alternate version](https://www.w3.org/TR/WCAG21/#dfn-conforming-alternate-version) that is the page itself. The read-aloud function **does** seem to violate the second conformance requirement [5.2.2 Full pages](https://www.w3.org/TR/WCAG21/#x5-2-2-full-pages) which states: "For the purpose of determining conformance, alternatives to part of a page's content are considered part of the page when the alternatives can be obtained directly from the page, e.g., a long description or an alternative presentation of a video." That is clearly the case, so from a strict reading perspective I guess the read-aloud function must fail the page.

**Comment by patrickhlauke** [comment]
> it feels so wrong to fail the site for this add-on content which hardly gets in the way

ultimately though, a site is responsible for third-party stuff they put on there, but i see your point. it does sound like in this very specific case (if it's just about the read-aloud, and it's not one of those overlays that then pretends to "fix" accessibility issues by providing text resizing, link underlining, etc) i could see an argument that it provides a (non-conforming) alternative version, in effect. it would probably be worth noting/raising as part of the rationale of an accessibility statement, or in the report itself, that issues found in it were excluded from the overall pass/fail assessment ... explaining the reason why failures in that widget/functionality were not taken into account when determining pass/fail for the page as a whole.

(and of course, would still be good to nonetheless note the problems with the functionality somewhere, even only as a best practice issue, maybe even to pass on to whoever develops that widget, to sort their rubbish out)

**Comment by detlevhfischer** [comment]
@mraccess77 
> WCAG allows for alternatives on the same page. An inaccessible graph could have a data table on the same page or linked to from that page to meet WCAG.

Just out of curiosity: Where does WCAG allow that, in the normative text? Or has there been a WG resolution at some point? The language around [conforming alternate version](https://www.w3.org/TR/WCAG21/#dfn-conforming-alternate-version) seems to suggest that the non-conforming version is another page (or at any rate, outside the conforming page). If non-conforming content is embedded on the same page, how can that be reconciled with Conformance requirement [5.2.2 Full pages](https://www.w3.org/TR/WCAG21/#x5-2-2-full-pages): "Conformance (and conformance level) is for full Web page(s) only, and cannot be achieved if part of a Web page is excluded"?

**Comment by bruce-usab** [comment]
> Where does WCAG allow that, in the normative text? Or has there been a WG resolution at some point? The language around conforming alternate version seems to suggest that the non-conforming version is another page (or at any rate, outside the conforming page).

The WG has had significant discussion on this point, but I do not believe it is well documented.  I concur that the normative text suggests other pages.  As I recall, the most significant conversation was around date-pickers where there is a mouse-only calendar view paired free-form text.  The consensus conclusion was that the normative text does not forbid the CAV being provided on the same page as the non-conforming content.

@alastc might be able to point to a WG resolution.  I find his ability to find those almost magical!

**Comment by mraccess77** [comment]
Any discussion of datepickers alternatives must also take into account whether entering a date manually actually provides the same functionality as a date picker - in many cases I'm not sure it does.

**Comment by patrickhlauke** [comment]
it's a bit tangential to this, but:

> in many cases I'm not sure it does

indeed, if the datepicker shows extra info (like days that can't be picked because there is no availability, or the store or whatever is closed, etc) then it's not equivalent unless there's a similar/equivalent static calendar view readily available.

**Comment by alastc** [comment]
Thanks @bruce-usab, but that ability is mostly for things since I've been a chair! With a little google-fu I could find this [old version of the conforming alt versions](https://www.w3.org/WAI/GL/2007/05/alternate-versions.html).

My reading of the [current conformance understanding](https://www.w3.org/WAI/WCAG21/Understanding/conformance#conforming-alt-versions) makes me think the original example is ok. Also, with technique [G190](https://www.w3.org/WAI/WCAG21/Techniques/general/G190), although it focuses on having a link to a alt version, I don't see why we'd penalize a page for having both versions on the _same_ page.

**Comment by bruce-usab** [comment]
Coming back to the original question, a read-aloud feature is presumably not the conforming version.  So, for example, the feature would not expect to be captioned ([SC 1.2.2](https://www.w3.org/TR/WCAG21/#captions-prerecorded)).  The feature must certainly not interfere with keyboard navigation of the page ([CR5](https://www.w3.org/TR/WCAG21/#cc5)).  I would certainly _expect_ the feature to keyboard accessible ((SC 2.1.1)[https://www.w3.org/TR/WCAG21/#keyboard] -- especially considering the range of people who would find the feature useful for.

**Comment by fstrr** [comment]
This was initially tagged as a discussion, so I'm going to move into the discussions area. Please close if the discussion is complete.
